rpc, doc: getpeerinfo updates

jonatack opened the pull request:
Various updates and fixups, mostly targeting v24. Please refer to the commit messages for details.

jonatack committed:
Update getpeerinfo/-netinfo/TxRelay#m_relay_txs relaytxes docs (for v24 backport)

to the current p2p behavior.  We only initialize the Peer::TxRelay m_relay_txs
data structure if it isn't an outbound block-relay-only connection and fRelay=true
(the peer wishes to receive tx announcements) or we're offering NODE_BLOOM to this peer.

jonatack committed:
Improve getpeerinfo pingtime, minping, and pingwait help docs

jonatack committed:
Make getpeerinfo field order consistent with its help (for v24 backport)

This also keeps it consistent with the last release (v23)

jonatack committed:
Always return getpeerinfo "minfeefilter" field (for v24 backport)

with its pre-existing v23 default value of 0.

MarcoFalke commented on src/rpc/net.cpp:
```diff
@@ -200,6 +200,7 @@ static RPCHelpMan getpeerinfo()
         ServiceFlags services{fStateStats ? statestats.their_services : ServiceFlags::NODE_NONE};
         obj.pushKV("services", strprintf("%016x", services));
         obj.pushKV("servicesnames", GetServicesNames(services));
+        obj.pushKV("relaytxes", fStateStats ? statestats.m_relay_txs : false);
```
A user reading `false` with bloom filters disabled may register the peer as one that txs aren't sent to, as the relay flag can't change after the handshake. So I think keeping this optional may be better. Also, in bitcoin-cli it might be better to display it as optional instead of a picking a constant which may later turn out to be wrong.

jonatack replied:
> keeping this optional may be better

Note that this patch maintains the current API, whereas leaving it optional would annoyingly break user space without notice in v24. The API change was unclear but it appears unintended as it added neither optional mentions in the getpeerinfo help, which you fixed afterward, nor a release note. Absent a real need to change it (say, like removing a field), I have the impression that users prefer we don't break things willy-nilly, at least without some explanation of what and why (if we really, really want to make it optional ;).

MarcoFalke replied:
I think whether the change was "unintended" was discussed already, so I am not going to reply to that.

If there is a missing release note, it can be added to the wiki.

jonatack replied:
To argue against my points, maybe getpeerinfo is mainly consumed by us developers and rarely by software clients or node operators -- I don't know -- and I agree with you that for us developers it is preferable not to return a relaytxes value while the connection is setting up. 

Updating to take your feedback (appreciated, thank you) with the caveats I described.

jonatack replied:
> If there is a missing release note, it can be added to the wiki.

Note that it doesn't seem to have been added to the wiki if I'm parsing it correctly.

jonatack replied:
I've also heard from the financial software industry using bitcoin that these v24 changes will break their production systems without a patch. It was first reported to me orally by an engineer who observed the discussions in #25923, and today I received written confirmation of this by PMs and engineers there.

jonatack commented:
Updated per @MarcoFalke's feedback (thanks!).

<details><summary><code>git diff bd56f68 e2f5c29</code></summary><p>

```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index 04d72ade3d..a8293afdc7 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -467,8 +467,8 @@ public:
         if (!batch[ID_NETWORKINFO]["error"].isNull()) return batch[ID_NETWORKINFO];
 
         const UniValue& networkinfo{batch[ID_NETWORKINFO]["result"]};
-        if (networkinfo["version"].getInt<int>() < 220000) {
-            throw std::runtime_error("This version of -netinfo requires bitcoind server to be running v22.0 and up");
+        if (networkinfo["version"].getInt<int>() < 239900) {
+            throw std::runtime_error("This version of -netinfo requires the bitcoind server to be running v24.0 and up");
         }
         const int64_t time_now{TicksSinceEpoch<std::chrono::seconds>(CliClock::now())};
 
@@ -478,7 +478,7 @@ public:
             const int8_t network_id{NetworkStringToId(network)};
             if (network_id == UNKNOWN_NETWORK) continue;
             const bool is_outbound{!peer["inbound"].get_bool()};
-            const bool is_tx_relay{peer["relaytxes"].get_bool()};
+            const bool is_tx_relay{peer["relaytxes"].isNull() ? true : peer["relaytxes"].get_bool()};
             const std::string conn_type{peer["connection_type"].get_str()};
             ++m_counts.at(is_outbound).at(network_id);      // in/out by network
             ++m_counts.at(is_outbound).at(NETWORKS.size()); // in/out overall
@@ -642,7 +642,7 @@ public:
         "  send     Time since last message sent to the peer, in seconds\n"
         "  recv     Time since last message received from the peer, in seconds\n"
         "  txn      Time since last novel transaction received from the peer and accepted into our mempool, in minutes\n"
-        "           \"*\" - whether we currently relay transactions to this peer (relaytxes is false)\n"
+        "           \"*\" - whether we relay transactions to this peer (relaytxes is false)\n"
         "  blk      Time since last novel block passing initial validity checks received from the peer, in minutes\n"
         "  hb       High-bandwidth BIP152 compact block relay\n"
         "           \".\" (to)   - we selected the peer as a high-bandwidth peer\n"
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 781debd132..28148e1c65 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -271,11 +271,11 @@ struct Peer {
 
     struct TxRelay {
         mutable RecursiveMutex m_bloom_filter_mutex;
-        /** Whether we currently relay transactions to this peer.
+        /** Whether we relay transactions to this peer.
          *
          * On receiving a `version` message, this is initially set to false
-         * either if the peer is outbound block-relay-only, or otherwise based
-         * on the fRelay flag.
+         * if the peer is outbound block-relay-only and otherwise based on
+         * the fRelay flag.
          *
          * If initially set to false, it can only be flipped to true if we have
          * offered the peer NODE_BLOOM services and it sends us a `filterload`
diff --git a/src/rpc/net.cpp b/src/rpc/net.cpp
index 41c4a2c539..e953f8ba25 100644
--- a/src/rpc/net.cpp
+++ b/src/rpc/net.cpp
@@ -114,7 +114,7 @@ static RPCHelpMan getpeerinfo()
                     {
                         {RPCResult::Type::STR, "SERVICE_NAME", "the service name if it is recognised"}
                     }},
-                    {RPCResult::Type::BOOL, "relaytxes", "Whether we currently relay transactions to this peer"},
+                    {RPCResult::Type::BOOL, "relaytxes", /*optional=*/true, "Whether we relay transactions to this peer"},
                     {RPCResult::Type::NUM_TIME, "lastsend", "The " + UNIX_EPOCH_TIME + " of the last send"},
                     {RPCResult::Type::NUM_TIME, "lastrecv", "The " + UNIX_EPOCH_TIME + " of the last receive"},
                     {RPCResult::Type::NUM_TIME, "last_transaction", "The " + UNIX_EPOCH_TIME + " of the last valid transaction received from this peer"},
@@ -200,7 +200,9 @@ static RPCHelpMan getpeerinfo()
         ServiceFlags services{fStateStats ? statestats.their_services : ServiceFlags::NODE_NONE};
         obj.pushKV("services", strprintf("%016x", services));
         obj.pushKV("servicesnames", GetServicesNames(services));
-        obj.pushKV("relaytxes", fStateStats ? statestats.m_relay_txs : false);
+        if (fStateStats) {
+            obj.pushKV("relaytxes", statestats.m_relay_txs);
+        }
         obj.pushKV("lastsend", count_seconds(stats.m_last_send));
         obj.pushKV("lastrecv", count_seconds(stats.m_last_recv));
         obj.pushKV("last_transaction", count_seconds(stats.m_last_tx_time));
```
</p></details>

mzumsande commented on src/net_processing.cpp:
```diff
@@ -271,12 +271,15 @@ struct Peer {
-         * us a `filterload` or `filterclear` message. See BIP37. */
+         * On receiving a `version` message, this is initially set to false
+         * if the peer is outbound block-relay-only and otherwise based on
+         * the fRelay flag.
```
I think that if this is an outbound block-relay-only peer, we don't create a `TxRelay` object for this peer in the first place, so  `TxRelay::m_relay_txs` is neither true nor false but doesn't exist.

jonatack replied:
Thanks, updated. I agree this was previously out of date and the proposed update was imprecise. It is already documented in the `PeerManagerImpl::ProcessMessage()::VERSION` code so it seems best to remove it here instead of duplicating.

mzumsande commented on src/net_processing.cpp:
```diff
@@ -271,12 +271,15 @@ struct Peer {
+         *
+         * If initially set to false, it can only be flipped to true if we have
+         * offered the peer NODE_BLOOM services and it sends us a `filterload`
+         * or `filterclear` message. See BIP37. */
```
This sentence could need an update after #22778: If we haven't offered `NODE_BLOOM` and get `fRelay=false`, we no longer set `m_relay_txs` to false, but now don't initialise a TxRelay object anymore.
So the case where it is initialised to `false` and is impossible to flip cannot happen anymore.

jonatack replied:
Thanks, removed per https://github.com/bitcoin/bitcoin/pull/26109/files#r975018566.

luke-jr commented on src/bitcoin-cli.cpp:
```diff
@@ -467,8 +467,8 @@ class NetinfoRequestHandler : public BaseRequestHandler
-        if (networkinfo["version"].getInt<int>() < 209900) {
-            throw std::runtime_error("-netinfo requires bitcoind server to be running v0.21.0 and up");
+        if (networkinfo["version"].getInt<int>() < 239900) {
+            throw std::runtime_error("This version of -netinfo requires the bitcoind server to be running v24.0 and up");
```
```suggestion
            throw std::runtime_error("This version of -netinfo requires the bitcoind server to be running v24.0 or newer");
```

jonatack replied:
Done

jonatack commented:
Updated per @mzumsande and @luke-jr review feedback (thanks).

MarcoFalke commented on src/rpc/net.cpp:
```diff
@@ -146,7 +146,7 @@ static RPCHelpMan getpeerinfo()
                         {RPCResult::Type::STR, "permission_type", Join(NET_PERMISSIONS_DOC, ",\n") + ".\n"},
                     }},
-                    {RPCResult::Type::NUM, "minfeefilter", /*optional=*/true, "The minimum fee rate for transactions this peer accepts"},
+                    {RPCResult::Type::NUM, "minfeefilter", "The minimum fee rate for transactions this peer accepts"},
```
Is is worth it to change the interface, given that there are other optional fields?

jonatack replied:
It maintains the existing interface in releases and follows your comment, "there are some fields that can obviously be made non-optional (such as the fee filter), because they can be absent on the wire even long after the connection has been spun up (thus require a default value anyway)".

jonatack replied:
(Anything that can have an always-present default is simpler for us and for API clients to deal with, code-wise.)

jonatack commented:
Tag for 24.0?

fanquake commented on src/rpc/net.cpp:
```diff
@@ -146,7 +146,7 @@ static RPCHelpMan getpeerinfo()
                     {
                         {RPCResult::Type::STR, "permission_type", Join(NET_PERMISSIONS_DOC, ",\n") + ".\n"},
                     }},
-                    {RPCResult::Type::NUM, "minfeefilter", /*optional=*/true, "The minimum fee rate for transactions this peer accepts"},
```
in 37d5cdbaac6f84aff90942c03f83e094a09c7bde:

I see this is a partial revert of #24691, which was a follow up to #21160? This commit would be easier to follow without the previous move-only commit.

> This was inadvertently changed on master after the v23 branch-off.

It would be good to link to actual discussion / history rather than saying "inadvertently changed". It's not clear that anything inadvertent happened. The change was made, docs/optional was added / fixed, but now it's being changed back again.

jonatack replied:
See the discussion from https://github.com/bitcoin/bitcoin/pull/26109#discussion_r973104660. It's hard to tell what happened and inadvertent seems the kindest explanation if one is needed, along with not linking.  If any fields become newly optional in v24, ISTM that should warned about in a release note (that users can be pointed to, if they ask what happened).

fanquake commented on src/rpc/net.cpp:
```diff
@@ -125,7 +125,7 @@ static RPCHelpMan getpeerinfo()
                     {RPCResult::Type::NUM, "pingtime", /*optional=*/true, "ping time (if available)"},
                     {RPCResult::Type::NUM, "minping", /*optional=*/true, "minimum observed ping time (if any at all)"},
-                    {RPCResult::Type::NUM, "pingwait", /*optional=*/true, "ping wait (if non-zero)"},
+                    {RPCResult::Type::NUM, "pingwait", /*optional=*/true, "The duration in milliseconds (ms) of an outstanding ping (if non-zero)"},
```
In 3c196b1c1dacc0c1d67d6c3ebd756e2614fef89f:

Doc change seems fine, but isn't a bugfix/backport right? Any reason to change just this field? i.e why not make the same change to `pingtime`?

jonatack replied:
> In [3c196b1](https://github.com/bitcoin/bitcoin/commit/3c196b1c1dacc0c1d67d6c3ebd756e2614fef89f):
> 
> Doc change seems fine, but isn't a bugfix/backport right? Any reason to change just this field? i.e why not make the same change to `pingtime`?

Good idea, will look at doing that. Agree, not a backport candidate.

jonatack replied:
Done as well for pingtime and minping in the same commit.

fanquake commented on src/bitcoin-cli.cpp:
```diff
@@ -467,8 +467,8 @@ class NetinfoRequestHandler : public BaseRequestHandler
         const UniValue& networkinfo{batch[ID_NETWORKINFO]["result"]};
-        if (networkinfo["version"].getInt<int>() < 209900) {
-            throw std::runtime_error("-netinfo requires bitcoind server to be running v0.21.0 and up");
+        if (networkinfo["version"].getInt<int>() < 239900) {
```
In e6603e50f4fbeeeb5a0dfd25f78b83b85d542984:
> This change is because the getpeerinfo "relaytxes" field was changed
between v23 and v24 to become optional

If we end up bumping/changing this anyway, because of the `relaytxes` change, do we actually need to revert `minfeefilter` to being non-optional?

jonatack replied:
The idea is to maintain the current released API interface and avoid potentially breaking user space without a compelling reason (`minfeefilter` was added to getpeerinfo in 2018 in https://github.com/bitcoin/bitcoin/commit/5778bf95d94f4d187451a72a5b86ba4daaafe4ab and I believe it has had the same behavior since then for around 7 releases. So far, `-netinfo` doesn't use it).

fanquake reviewed (commented):
I have trouble keeping up with bitcoin-cli / netinfo changes. I think in general us changing code / reverting things to try and maintain compatibility with `-netinfo` is somewhat backwards. It should really be a best-effort component of bitcoin-cli, that just adapts to any code changes we make, not another interface / tool we need to worry about backwards compatibility for.

jonatack commented:
> It [-netinfo] should really be a best-effort component of bitcoin-cli, that just adapts to any code changes we make

That's what is done here, as well as in #25176, for this release: adapting -netinfo to changes.

jonatack commented on src/rpc/net.cpp:
```diff
@@ -200,6 +200,9 @@ static RPCHelpMan getpeerinfo()
         ServiceFlags services{fStateStats ? statestats.their_services : ServiceFlags::NODE_NONE};
         obj.pushKV("services", strprintf("%016x", services));
         obj.pushKV("servicesnames", GetServicesNames(services));
+        if (fStateStats) {
```
> Can you clarify how this commit (standalone) is a bug fix that needs to be backported to 24.x? (what will break if this isn't done).

Other already-backported v24 commits aren't bug fixes; some are doc updates and some are even arguably features. 

The getpeerinfo RPC has many fields and they were in the same order in the help and the output in the previous releases. I don't see a compelling reason to break that for these two fields for v24 and needlessly make the RPC more confusing.

mzumsande replied:
JSON objects are by [definition](https://www.rfc-editor.org/rfc/rfc8259.html)  collections of name/value pairs that are unordered.
Given that, why not just shuffle the order in the help instead, so that we don't need extra conditionals in the source code and can group things together that belong together?

jonatack replied:
On my Debian and macOS machines, the CLI and its help returns the getpeerinfo fields in the same order, which the codebase handles in the same order as well. This is handy; I don't see a very compelling reason to change the API needlessly to be less helpful for those who read the help and run CLI commands.

jonatack replied:
> can group things together that belong together?

It would be potentially nice for that `fStateStats` flag grouping to go away; it has been a source of bugs in the past.

fanquake replied:
> to be less helpful for those who read the help and run CLI commands.

The point is that machines / parsers consuming the JSON do not, and should not care about ordering. Which means nothing is "broken" if the fields happen to change order, and modifying the code to preserve an order that isn't guaranteed by anything is pointless. We also aren't optimising our RPC output for humans reading JSON in a terminal.

> to change the API needlessly

The API isn't changing though. The ordering is not guaranteed, and is not part of any API. The fields returned remain exactly the same.

I agree with Martins comment, if you care about the ordering being the same, it would make more sense to change the docs.

jonatack replied:
Our RPC interface has the dual constraint of providing an API consumed by machines and the CLI interface used by humans -- cf the RPC helps that provide at least one example of each. It is therefore preferable to maintain the order, particularly when there is no need to change it.

Diffing `src/rpc/net.cpp` between the previous releases and master shows that changing the help order instead of the field order would mean needless churn in the interface. Needless, because there is no compelling reason. It makes more sense to keep the help the same in the absence of a real need to change it. A change ought to be justified, not stability. This commit chooses stability.

fanquake replied:
>  shows that changing the help order instead of the field order would mean needless churn in the interface.

The documentation is not an interface, and changing the order of fields listed within it does not change the RPC interface/API. Ideally, the docs would be auto-generated from the code, and then more than likely, the fields would just be ordered alphanumerically.

> Needless, because there is no compelling reason. It makes more sense to keep the help the same in the absence of a real need to change it. A change from release to release ought to be justified, not stability. This commit chooses stability.

I disagree that the solution to maintaining something (field ordering) that isn't guaranteed, or part of any API, is [to add more complexity too](https://github.com/bitcoin/bitcoin/pull/26109#discussion_r977743011), or change the code. Especially when you could change the documentation and acheive the same thing.

jonatack replied:
The diff to maintain stability is essentially the same without needlessly changing field position for the user. Code serves users.

jonatack replied:
(I would like to remove the `fStateStats` grouping completely for the next release.)

jonatack replied:
> Ideally the fields would just be ordered alphanumerically

That would be of benefit to neither machine nor humans? Rather it would be needless churn and confusion for humans reading the docs and using the RPC via the command line.

At least some of the RPCs return some of their output in an order that makes it easier to understand for a human, i.e. getpeerinfo begins with id/address/network fields for each peer. This seems helpful.

Whether everyone really wants to willy-nilly reorder all the fields and help docs across the board to make life confusing for people who use the CLI or not, it's a little strange to go way out on that limb in reaction to a minimal fixup.

fanquake replied:
> That would be of benefit to neither machine nor humans?

It would benefit both (and developers) by:
* providing an ordering that is no-longer arbitrary
* which can be tested/linted/enforced
* same mechanism can be used to produce the field output, and the help output.
* we can stop caring about / trying to enforce / fix completely arbitrary orderings, that are not enforced / tested / checked by anything.

jonatack replied:
Machines don't care about ordering and humans can find ordering helpful.  

Such a large change won't be done for v24 (if at all) and the argument isn't relevant here.

jonatack commented:
Updated the commit messages to indicate proposed backport (and removed mentions of previous inadvertent changes), and took @fanquake's feedback to improve the help documentation for the `pingtime` and `minping` fields in [a3789c7](https://github.com/bitcoin/bitcoin/pull/26109/commits/a3789c700b5a43efd4b366b4241ae840d63f2349).

luke-jr commented on src/bitcoin-cli.cpp:
```diff
@@ -467,8 +467,8 @@ class NetinfoRequestHandler : public BaseRequestHandler
         const UniValue& networkinfo{batch[ID_NETWORKINFO]["result"]};
-        if (networkinfo["version"].getInt<int>() < 209900) {
-            throw std::runtime_error("-netinfo requires bitcoind server to be running v0.21.0 and up");
+        if (networkinfo["version"].getInt<int>() < 239900) {
```
The commit message appears to have lost the rationale for this bump.

The reason in e6603e50f4fbeeeb5a0dfd25f78b83b85d542984 also doesn't make sense: making a field optional should not break compatibility.

luke-jr replied:
(I'd also prefer inlining the rationale by checking the presence of the feature itself, rather than a numeric version comparison...)

jonatack replied:
@luke-jr see #25176 for the patch required that will only be present in the next release.

jonatack replied:
> (I'd also prefer inlining the rationale by checking the presence of the feature itself, rather than a numeric version comparison...)

Multiple fields could change, so it's simpler to check the server version.

luke-jr replied:
@jonatack #25176 is backward compatible. Only the server side broke compatibility, and fixing it on the client side shouldn't affect operation with older nodes.

> Multiple fields could change, so it's simpler to check the server version.

Only the first one encountered that would cause an error needs to be checked. This is just a "friendlier error message", right?

jonatack replied:
Ah, you're right. It should instead check for minimum v22 when the 3 `addr_*` fields were added. Your suggestion would be to check if one of them IsNull()? (Yes, a friendlier message, it replaces the "JSON value is not of expected type" errors.)

luke-jr replied:
Yes, I'd just remove it temporarily to see where it fails first, and add a check at that location.

jonatack replied:
Done, thanks.

jonatack replied:
Thanks again @luke-jr. I've dropped that last commit and removed "-netinfo" from the PR title and description. After checking back version by version, the -netinfo code always handled the absence or presence of the three addr_* fields in a backward compatible manner, so no need to update -netinfo further for v24.

luke-jr commented on src/rpc/net.cpp:
```diff
@@ -246,6 +247,7 @@ static RPCHelpMan getpeerinfo()
             permissions.push_back(permission);
         }
         obj.pushKV("permissions", permissions);
+        obj.pushKV("minfeefilter", fStateStats ? ValueFromAmount(statestats.m_fee_filter_received) : 0);
```
This seems broken. It's not necessarily 0 if `fStateStats` is unavailable...?

jonatack replied:
It has been the default since the field was added in 2018 in https://github.com/bitcoin/bitcoin/commit/5778bf95d94f4d187451a72a5b86ba4daaafe4ab. This only maintains that behavior.

luke-jr replied:
Returning the default here is incorrect. It may be set for the peer - we don't know.

jonatack replied:
Recent review discussion seems in favor of keeping the minfeefilter behavior with a default value.

- https://github.com/bitcoin/bitcoin/pull/25923#pullrequestreview-1086617087: "there are some fields that can obviously be made non-optional (such as the fee filter), because they can be absent on the wire even long after the connection has been spun up (thus require a default value anyway)"

- https://github.com/bitcoin/bitcoin/pull/25923#issuecomment-1227233657: "I think it is ok to display the values relaytxes=false and minfeefilter=0...this is what the code did before PR 21160. If this is stripped to the bare minimum needed for the fix, I think it will have higher chance of making it to 24.0"

Though perhaps with a different default value (not proposed here).

- https://github.com/bitcoin/bitcoin/pull/25923#issuecomment-1233070570: "I could see -1 or CAmount::max() being better defaults for minfeefilter if relaying isn't happening though, since that way you could interpret it correctly without needing to also look at relaytxes"

brunoerg reviewed (approved):
ACK a3789c700b5a43efd4b366b4241ae840d63f2349

In 9cd6682545e845277d2207654250d1ed78d0c695, I just compared the results (master x this branch) and could realize now it's consistent with its help.

In 9cd6682545e845277d2207654250d1ed78d0c695, I modified the code to test it, like:
```diff
diff --git a/src/rpc/net.cpp b/src/rpc/net.cpp
index 704a23240..9acee54c6 100644
--- a/src/rpc/net.cpp
+++ b/src/rpc/net.cpp
@@ -184,7 +184,7 @@ static RPCHelpMan getpeerinfo()
     for (const CNodeStats& stats : vstats) {
         UniValue obj(UniValue::VOBJ);
         CNodeStateStats statestats;
-        bool fStateStats = peerman.GetNodeStateStats(stats.nodeid, statestats);
+        bool fStateStats = false;
         obj.pushKV("id", stats.nodeid);
         obj.pushKV("addr", stats.m_addr_name);
         if (stats.addrBind.IsValid()) {
```

So, I ran `getpeerinfo`, and I could realize `minfeefilter` has been returned even with `fStateStats` false, e.g.:
```
  {
    "id": 8,
    "addr": "34.121.41.161:8333",
    "addrbind": "192.168.0.73:49948",
    "network": "ipv4",
    "services": "0000000000000000",
    "servicesnames": [
    ],
    "lastsend": 1664375138,
    "lastrecv": 1664375138,
    "last_transaction": 0,
    "last_block": 0,
    "bytessent": 127,
    "bytesrecv": 198,
    "conntime": 1664375138,
    "timeoffset": 0,
    "version": 0,
    "subver": "",
    "inbound": false,
    "bip152_hb_to": false,
    "bip152_hb_from": false,
    "permissions": [
    ],
    "minfeefilter": 0,
    "bytessent_per_msg": {
      "version": 127
    },
    "bytesrecv_per_msg": {
      "sendaddrv2": 24,
      "verack": 24,
      "version": 126,
      "wtxidrelay": 24
    },
    "connection_type": "outbound-full-relay"
  }
```

While reviewing it, it seemed a little bit weird to return `"minfeefilter": 0` because - as a user - I wouldn't be able to know if it returned 0 because it is the real value from `m_fee_filter_received` or because `fStateStats` is false. However, reading some discussions (https://github.com/bitcoin/bitcoin/pull/25923#pullrequestreview-1086617087 and https://github.com/bitcoin/bitcoin/pull/25923#issuecomment-1227233657) the reason seems ok for me, but I'd prefer to return -1 as proposed by https://github.com/bitcoin/bitcoin/pull/25923#issuecomment-1233070570 because it would be easier to distinguish that there aren't any stats.

df660ddb1cce1ee330346fe1728d868f41ad0256 and a3789c700b5a43efd4b366b4241ae840d63f2349 are both docs improvement, they seem good for me.

jonatack commented:
Thanks for the review @brunoerg!

> I'd prefer to return -1 as proposed by [#25923 (comment)](https://github.com/bitcoin/bitcoin/pull/25923#issuecomment-1233070570) because it would be easier to distinguish that there aren't any stats.

Yes, I don't disagree but left the default as it has been since the field was added in 2018 as the commit is intended for backport, unless people feel it's worth including.

@MarcoFalke mind re-acking your review approval?

fanquake reviewed (commented):
I still don't understand 9cd6682545e845277d2207654250d1ed78d0c695, and it specifically being marked "for v24 backport". We don't make code changes in master primarily to backport them, we make changes in master (ideally) to fix bugs, and then, backport (select) bug fixes as required.

9cd6682545e845277d2207654250d1ed78d0c695 does not fix a bug. It moves around code, while making it more complex (additional conditionals, ungroups related items), to make JSON output align with help documentation, which as far as I'm aware, is neither a goal we have across RPCs generally, nor enforced by anything. [JSON](https://www.json.org/json-en.html) does not have field ordering, so changing our code to maintain field ordering, doesn't make sense.

If we really wanted them to align, an alternative would be to change the help docs to reflact the new ordering, rather than make the code worse.

This was discussed at length in [this thread](https://github.com/bitcoin/bitcoin/pull/26109/files#r977380363).

jonatack commented:
The field and help went out of sync between v23 and now, so it makes sense both for master and for v24.

I've been asked in the past to mention "for backport" in the commit names, but of course can drop that mention if that is now no longer our preferred practice.

> we make changes in master (ideally) to fix bugs, and then, backport (select) bug fixes as required... [9cd6682](https://github.com/bitcoin/bitcoin/commit/9cd6682545e845277d2207654250d1ed78d0c695) does not fix a bug.

You have tagged non-bugfixes for backport to v24. For instance: https://github.com/bitcoin/bitcoin/pull/26145/commits/68209a7b5c0326e14508d9cf749771605bd6ffe7. The other commit in that PR wasn't a bug, either.

> [JSON](https://www.json.org/json-en.html) does not have field ordering

I've addressed this with https://github.com/bitcoin/bitcoin/pull/26109#discussion_r979843696: *"Our RPC interface has the dual constraint of providing an API consumed by machines and the CLI interface used by humans -- cf the RPC helps that provide at least one example of each. It is therefore preferable to maintain the order, particularly when there is no need to change it."*

> If we really wanted them to align

Of course we do! If you look at the help and result of our RPCs --- including those with many fields like getpeerinfo, getnetworkinfo, getdeploymentinfo, etc. --- the help and the result are in the same order. This isn't a coincidence.

> an alternative would be to change the help docs to reflact the new ordering

There is no need or reason to change the ordering in the first place. This represents the least change for the user, which is more worthwhile than saving a line of code, or at least not worth the hostility you've been showing here.

vasild reviewed (approved):
ACK a3789c700b5a43efd4b366b4241ae840d63f2349

I agree changing order of JSON fields is not a change or breaking of the API. Anything that breaks due to that is broken already. However, I sometimes consume the JSON output with my eyes. It would be nice if the order does not change (it follows that I am semi-broken). Also, it would be nice if the order in the `cpp` file of the help and the place where fields are "printed" is the same (may help devs a little bit when navigating the source code).

Non-blocker: I think the versions mentions do not belong to the commit messages, they should be timeless. Imagine if a commit like "... previous version v23 ..." is merged sometime in the future when the previous version is not 23 anymore. Also, "for backport in v24" implies that the decision is made to backport it already. Commits should make sense on their own, regardless of when releases are done. The decision whether to backport is a separate one.

jonatack commented:
> Non-blocker: I think the versions mentions do not belong to the commit messages

Thanks! Yes, I've been asked by a maintainer to add backport mentions to commit messages for previous releases. IDK if current practices have changed.

achow101 commented:
ACK a3789c700b5a43efd4b366b4241ae840d63f2349

I agree with @vasild's non-blocking comments, however this is small and simple enough to go forward with at this point.

I do think the RPC changes are backportable to 24.0 as corrections to user facing documentation is important to backport.

MarcoFalke commented on src/bitcoin-cli.cpp:
```diff
@@ -642,7 +642,7 @@ class NetinfoRequestHandler : public BaseRequestHandler
         "  recv     Time since last message received from the peer, in seconds\n"
         "  txn      Time since last novel transaction received from the peer and accepted into our mempool, in minutes\n"
-        "           \"*\" - the peer requested we not relay transactions to it (relaytxes is false)\n"
+        "           \"*\" - whether we relay transactions to this peer (relaytxes is false)\n"
```
This doesn't seem right. Even if relaytxes is true, this might be false (for example this is an outbound block-relay-only connection?)

MarcoFalke replied:
unrelated: It seems the cli will default to true if the field is not available, which might also be confusing if an outbound block-relay-only connection is spun up.

jonatack replied:
Good catch. Perhaps: `we don't relay transactions to this peer (relaytxes is false or not available)`

jonatack replied:
Proposed a fixup in #26328.

MarcoFalke replied:
Oh, I think I read this as `version.relaytxes` is false, but this means `getpeerinfo.relaytxes` is false? The two are different, so for one interpretation this comment is right and for the other it is wrong?

jonatack replied:
Will clarify next time I touch this that it's the `getpeerinfo "relaytxes" field` (what is `version.relaytxes`?)

MarcoFalke replied:
> what is version.relaytxes?

https://github.com/bitcoin/bips/blob/master/bip-0037.mediawiki#extensions-to-existing-messages

jonatack replied:
> [bitcoin/bips@`master`/bip-0037.mediawiki#extensions-to-existing-messages](https://github.com/bitcoin/bips/blob/master/bip-0037.mediawiki?rgh-link-date=2023-01-11T17%3A12%3A10Z#extensions-to-existing-messages)

Thanks!

jonatack replied:
Thanks! The various namings for tx relay can be a little confusing.
